### PR TITLE
Avoid caching duplicate export info from AutoImportProvider

### PR DESCRIPTION
**Pull request by andrewbranch**
Noticed this while investigating #52660. I don’t think it’s related to the crash or even an observably buggy behavior, but this should save a bit of time and memory for almost everyone—hoping it shows up on server perf tests.

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot perf test this
@typescript-bot user test tsserver

**Comment by andrewbranch** [comment]
Running the change locally took the `getExportInfoMap` step from 60ms to 40ms.

File: src/services/exportInfoMap.ts [hunk 1]
```diff
@@ -422,7 +422,16 @@ export function forEachExternalModuleToImportFrom(
     const autoImportProvider = useAutoImportProvider && host.getPackageJsonAutoImportProvider?.();
     if (autoImportProvider) {
         const start = timestamp();
-        forEachExternalModule(autoImportProvider.getTypeChecker(), autoImportProvider.getSourceFiles(), excludePatterns, (module, file) => cb(module, file, autoImportProvider, /*isFromPackageJson*/ true));
+        const checker = program.getTypeChecker();
+        forEachExternalModule(autoImportProvider.getTypeChecker(), autoImportProvider.getSourceFiles(), excludePatterns, (module, file) => {
+            if (file && !program.getSourceFile(file.fileName) || !file && !checker.resolveName(module.name, /*location*/ undefined, SymbolFlags.Module, /*excludeGlobals*/ false)) {
```
Comment (DanielRosenwasser) [inline src/services/exportInfoMap.ts:427]: Out of curiosity, what does it mean for a module to be associated with no file? Ambient modules?

**Comment by jakebailey** [comment]
Unless they already exist, I think the tsserver errors above appear to be real?

**Comment by DanielRosenwasser** [comment]
They are, but are likely unrelated to this change. The other day I made the error detector start requesting `completionEntryDetails` on all auto-imports (https://github.com/microsoft/typescript-error-deltas/pull/95), and it caught a bunch of new issues (https://github.com/microsoft/TypeScript/issues/52660)

File: src/services/exportInfoMap.ts (hunk 1 above)
Comment (andrewbranch) [reply src/services/exportInfoMap.ts:427]: Yep

**Comment by andrewbranch** [comment]
Nothing shows up in the perf run, but it also doesn’t show memory usage for tsserver

**Comment by andrewbranch** [comment]
I tried two more example repos locally, and both showed a significant improvement. On my “slow auto imports” repo, `getExportInfoMap` had a 35% improvement for a 15% improvement in overall response time!

**Comment by jakebailey** [comment]
What is your "slow auto imports" repo? I want to add it to my shortlist of new benchmarks; I don't think we have great coverage of that area but I want to make sure we eventually do once we have increased benchmark capacity.

File: src/services/exportInfoMap.ts [hunk 2]
```diff
@@ -422,7 +422,16 @@ export function forEachExternalModuleToImportFrom(
     const autoImportProvider = useAutoImportProvider && host.getPackageJsonAutoImportProvider?.();
     if (autoImportProvider) {
         const start = timestamp();
-        forEachExternalModule(autoImportProvider.getTypeChecker(), autoImportProvider.getSourceFiles(), excludePatterns, (module, file) => cb(module, file, autoImportProvider, /*isFromPackageJson*/ true));
+        const checker = program.getTypeChecker();
```
Comment (jakebailey) [inline src/services/exportInfoMap.ts:425]: Is there a situation in which this checker is different than `autoImportProvider.getTypeChecker()`? Presumably, given I guess they're different Programs?

**Review by jakebailey — APPROVED** [review]

File: src/services/exportInfoMap.ts (hunk 2 above)
Comment (andrewbranch) [reply src/services/exportInfoMap.ts:425]: Yes, they’re always different programs/checkers.

**Comment by andrewbranch** [comment]
> What is your "slow auto imports" repo?

It’s not an actual repo (and it’s also not _that_ slow), it’s just

```
{
  "devDependencies": {
    "typescript": "^4.7.2"
  },
  "dependencies": {
    "@angular/core": "^14.0.0",
    "@angular/forms": "^14.0.0",
    "@reduxjs/toolkit": "^1.8.2",
    "@types/node": "^17.0.40",
    "@types/react": "^18.0.12",
    "antd": "^4.21.0",
    "aws-sdk": "^2.1148.0",
    "mobx": "^6.6.0",
    "rxjs": "^7.5.5",
    "xstate": "^4.32.1"
  }
}
```

with `"module": "nodenext"`

File: src/services/exportInfoMap.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/services/exportInfoMap.ts:427]: IF you are checking for ambient module is this more readable: to use `tryFindAmbientModule` or `tryFindAmbientModuleWithoutAugmentations`

**Pull request merged**